File: compiler-rust/benchmarks/clbg/c/spectralnorm.c
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Ledrug
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double A(int i, int j) {
    return 1.0 / ((i + j) * (i + j + 1) / 2 + i + 1);
}

void Av(int n, double *v, double *out) {
    for (int i = 0; i < n; i++) {
        double sum = 0;
        for (int j = 0; j < n; j++) {
            sum += A(i, j) * v[j];
        }
        out[i] = sum;
    }
}

void Atv(int n, double *v, double *out) {
    for (int i = 0; i < n; i++) {
        double sum = 0;
        for (int j = 0; j < n; j++) {
            sum += A(j, i) * v[j];
        }
        out[i] = sum;
    }
}

void AtAv(int n, double *v, double *out, double *tmp) {
    Av(n, v, tmp);
    Atv(n, tmp, out);
}

int main(int argc, char *argv[]) {
    int n = argc > 1 ? atoi(argv[1]) : 100;

    double *u = malloc(n * sizeof(double));
    double *v = malloc(n * sizeof(double));
    double *tmp = malloc(n * sizeof(double));

    for (int i = 0; i < n; i++) {
        u[i] = 1.0;
    }

    for (int i = 0; i < 10; i++) {
        AtAv(n, u, v, tmp);
        AtAv(n, v, u, tmp);
    }

    double vBv = 0, vv = 0;
    for (int i = 0; i < n; i++) {
        vBv += u[i] * v[i];
        vv += v[i] * v[i];
    }

    printf("%.9f\n", sqrt(vBv / vv));

    free(u);
    free(v);
    free(tmp);
    return 0;
}

File: compiler-rust/benchmarks/clbg/rust/spectralnorm.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// contributed by the Rust Project Developers

fn a(i: usize, j: usize) -> f64 {
    ((i + j) * (i + j + 1) / 2 + i + 1) as f64
}

fn av(n: usize, v: &[f64], out: &mut [f64]) {
    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum += v[j] / a(i, j);
        }
        out[i] = sum;
    }
}

fn atv(n: usize, v: &[f64], out: &mut [f64]) {
    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum += v[j] / a(j, i);
        }
        out[i] = sum;
    }
}

fn atav(n: usize, v: &[f64], out: &mut [f64], tmp: &mut [f64]) {
    av(n, v, tmp);
    atv(n, tmp, out);
}

fn main() {
    let n: usize = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(100);

    let mut u = vec![1.0; n];
    let mut v = vec![0.0; n];
    let mut tmp = vec![0.0; n];

    for _ in 0..10 {
        atav(n, &u, &mut v, &mut tmp);
        atav(n, &v, &mut u, &mut tmp);
    }

    let mut vbv = 0.0;
    let mut vv = 0.0;
    for i in 0..n {
        vbv += u[i] * v[i];
        vv += v[i] * v[i];
    }

    println!("{:.9}", (vbv / vv).sqrt());
}
